From 3a43f34c41cdfc9b0f235c6881f3d0f8d7d99bf0 Mon Sep 17 00:00:00 2001
From: Dylan Yip <dylan.yip@xilinx.com>
Date: Tue, 9 Mar 2021 12:36:44 -0800
Subject: [PATCH 1772/1851] drm: xlnx: zynqmp_disp: Add set_timing support for
 live bridge

commit 8acfecbbc864d7bba8b7c49ffbb8e53f7dbf26f9 from
https://github.com/Xilinx/linux-xlnx.git

Currently PS DP live bridge expects the PS DP CRTC to be used so that it
can obtain the video timing from the PS DP CRTC. However, if an external
CRTC is used, the PS DP CRTC state mode will be invalid. This patch adds
support for a set_timing function so an external CRTC can share the
video timing with the PS DP encoder.

Signed-off-by: Dylan Yip <dylan.yip@xilinx.com>
Signed-off-by: Jianqiang Chen <jianqiang.chen@xilinx.com>
Acked-by: Varunkumar Allagadapa <varunkumar.allagadapa@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/gpu/drm/xlnx/zynqmp_disp.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/drivers/gpu/drm/xlnx/zynqmp_disp.c b/drivers/gpu/drm/xlnx/zynqmp_disp.c
index ffd61635f055..74fc0c1d3885 100644
--- a/drivers/gpu/drm/xlnx/zynqmp_disp.c
+++ b/drivers/gpu/drm/xlnx/zynqmp_disp.c
@@ -3125,6 +3125,19 @@ static int zynqmp_disp_bridge_get_input_fmts(struct xlnx_bridge *bridge,
 	return 0;
 }
 
+static int zynqmp_disp_bridge_set_timing(struct xlnx_bridge *bridge,
+					     struct videomode *vm)
+{
+	struct zynqmp_disp_layer *layer = bridge_to_layer(bridge);
+	struct zynqmp_disp *disp = layer->disp;
+	struct drm_crtc *crtc = &disp->xlnx_crtc.crtc;
+	struct drm_display_mode *adjusted_mode = &crtc->state->adjusted_mode;
+
+	drm_display_mode_from_videomode(vm, adjusted_mode);
+
+	return 0;
+}
+
 /*
  * Component functions
  */
@@ -3352,6 +3365,7 @@ int zynqmp_disp_probe(struct platform_device *pdev)
 		layer->bridge.set_input = &zynqmp_disp_bridge_set_input;
 		layer->bridge.get_input_fmts =
 			&zynqmp_disp_bridge_get_input_fmts;
+		layer->bridge.set_timing = &zynqmp_disp_bridge_set_timing;
 		layer->bridge.of_node = layer->of_node;
 		xlnx_bridge_register(&layer->bridge);
 	}
-- 
2.31.1

